File: pkg-r/R/PinSource.R
#' Pin Source
#'
#' @description
#' A DataSource implementation that reads data from a
#' [pins](https://pins.rstudio.com/) board. When the `"duckdb"` engine is used
#' and the pin type is one DuckDB can read natively (parquet, CSV, JSON), the
#' data is loaded directly from the cached pin files into DuckDB without
#' deserializing into R. For other pin types (e.g. RDS), or when the `"sqlite"`
#' engine is used, the data is deserialized via `pin_read()` and must produce
#' a data frame (or tibble), which is then registered with the chosen engine
#' just like [DataFrameSource].
#'
#' When loaded into DuckDB, the connection's external file access is locked
#' down so that LLM-generated SQL cannot reach the filesystem.
#'
#' If the pin has a title, description, or tags, [QueryChat] uses them as
#' the default `data_description`, which you can override.
#'
#' @section Lazy queries with pins:
#'
#' `PinSource` materializes the full dataset into DuckDB. For large parquet
#' pins where you want lazy query execution, read the pin files yourself and
#' pass a `tbl_sql` to [querychat()] instead:
#'
#' ```r
#' paths <- pins::pin_download(board, "my_pin")
#' con <- DBI::dbConnect(duckdb::duckdb())
#' DBI::dbExecute(
#'   con,
#'   sprintf("CREATE VIEW my_pin AS SELECT * FROM read_parquet('%s')", paths[1])
#' )
#' qc <- querychat(dplyr::tbl(con, "my_pin"))
#' ```
#'
#' The pin files are still downloaded to a local cache --- `pin_download()`
#' always fetches them. But rather than loading everything into memory, DuckDB
#' reads the parquet file lazily through dbplyr.
#'
#' This approach skips the security lockdown that `PinSource` applies, so
#' LLM-generated SQL can access files on the local system.
#'
#' @examples
#' if (rlang::is_installed(c("pins", "duckdb"))) {
#'   # Create a temporary board and pin some data
#'   board <- pins::board_temp()
#'   pins::pin_write(board, mtcars, "mtcars", type = "parquet")
#'
#'   # Create a PinSource
#'   ps <- PinSource$new(board, "mtcars")
#'
#'   # Query the pinned data
#'   ps$execute_query("SELECT * FROM mtcars WHERE mpg > 25")
#'
#'   ps$cleanup()
#' }
#'
#' @export
PinSource <- R6::R6Class(
  "PinSource",
  inherit = DBISource,
  public = list(
    #' @description
    #' Create a new PinSource
    #'
    #' @param board A pins board object (e.g. from [pins::board_folder()] or
    #'   [pins::board_connect()]).
    #' @param name Name of the pin to read.
    #' @param ... Not used; included for extensibility.
    #' @param table_name Name to use for the table in SQL queries. Defaults to
    #'   the pin name.
    #' @param version Pin version to read. If `NULL` (default), reads the
    #'   latest version.
    #' @param engine Database engine to use: `"duckdb"` or `"sqlite"`. Set the
    #'   global option `querychat.DataFrameSource.engine` to specify the default
    #'   engine. If `NULL` (default), uses the first available engine from
    #'   duckdb or RSQLite (in that order). Parquet, CSV, and JSON pins are read
    #'   most efficiently with the `"duckdb"` engine; with `"sqlite"` they are
    #'   deserialized via `pin_read()` instead.
    #'
    #' @return A new PinSource object
    initialize = function(
      board,
      name,
      ...,
      table_name = name,
      version = NULL,
      engine = getOption("querychat.DataFrameSource.engine", NULL)
    ) {
      rlang::check_installed("pins", reason = "to use PinSource.")
      rlang::check_dots_empty()

      engine <- engine %||% get_default_dataframe_engine()
      engine <- tolower(engine)
      arg_match(engine, c("duckdb", "sqlite"))

      table_name <- sanitize_table_name(table_name)
      private$.pin_meta <- pins::pin_meta(board, name, version = version)

      pin_type <- private$.pin_meta$type
      duckdb_file_types <- c("parquet", "csv", "json")
      use_duckdb_file_read <- engine == "duckdb" &&
        pin_type %in% duckdb_file_types

      if (use_duckdb_file_read) {
        check_installed("duckdb")
        con <- DBI::dbConnect(duckdb::duckdb())
        con_owned <- FALSE
        on.exit(if (!con_owned) DBI::dbDisconnect(con), add = TRUE)

        paths <- pins::pin_download(board, name, version = version)
        if (length(paths) != 1) {
          cli::cli_abort(
            "Pin {.val {name}} contains {length(paths)} files, but PinSource requires a single-file pin (as created by {.fn pins::pin_write})."
          )
        }
        reader_fn <- switch(
          pin_type,
          parquet = "read_parquet",
          csv = "read_csv_auto",
          json = "read_json_auto"
        )
        if (pin_type == "json") {
          DBI::dbExecute(con, "INSTALL json")
          DBI::dbExecute(con, "LOAD json")
        }
        quoted_path <- DBI::dbQuoteLiteral(con, paths[[1]])
        sql <- sprintf(
          "CREATE TABLE %s AS SELECT * FROM %s(%s)",
          DBI::dbQuoteIdentifier(con, table_name),
          reader_fn,
          quoted_path
        )
        DBI::dbExecute(con, sql)
        duckdb_lock_down(con)
      } else {
        if (engine == "sqlite" && pin_type %in% duckdb_file_types) {
          cli::cli_warn(
            c(
              "Reading {pin_type} pin {.val {name}} into SQLite.",
              "i" = "The {.pkg duckdb} engine reads {pin_type} pins more efficiently. Install {.pkg duckdb} or pass {.code engine = \"duckdb\"} to read the pin files directly."
            )
          )
        }
        data <- pins::pin_read(board, name, version = version)
        if (!is.data.frame(data)) {
          cli::cli_abort(
            "Pin {.val {name}} contains {.obj_type_friendly {data}}, not a data frame."
          )
        }
        con <- new_dataframe_connection(data, table_name, engine)
        con_owned <- FALSE
        on.exit(if (!con_owned) DBI::dbDisconnect(con), add = TRUE)
      }

      super$initialize(con, table_name)
      con_owned <- TRUE
    },

    #' @description
    #' Get a human-readable description of the pin for use in the system prompt.
    #'
    #' @return A string with the pin title, description, and tags, or an empty
    #'   string if none are set.
    get_data_description = function() {
      meta <- private$.pin_meta
      parts <- character()
      if (nzchar(meta$title %||% "")) {
        parts <- c(parts, meta$title)
      }
      if (nzchar(meta$description %||% "")) {
        parts <- c(parts, meta$description)
      }
      if (length(meta$tags) > 0) {
        parts <- c(parts, paste("Tags:", paste(meta$tags, collapse = ", ")))
      }
      paste(parts, collapse = "\n\n")
    }
  ),
  private = list(
    .pin_meta = NULL
  )
)

File: pkg-py/src/querychat/_pin_source.py
from __future__ import annotations

import re
from typing import TYPE_CHECKING, Any, TypeGuard

import duckdb
import narwhals.stable.v1 as nw

from ._datasource import (
    DataSource,
    MissingColumnsError,
    duckdb_get_schema,
    duckdb_lock_down,
)
from ._utils import check_query

if TYPE_CHECKING:
    from pins.boards import BaseBoard


def is_pins_board(obj: object) -> TypeGuard[BaseBoard]:
    try:
        from pins.boards import BaseBoard

        return isinstance(obj, BaseBoard)
    except ImportError:
        return False

DUCKDB_FILE_TYPES = {"parquet", "csv", "json"}
DUCKDB_READER_FN = {
    "parquet": "read_parquet",
    "csv": "read_csv_auto",
    "json": "read_json_auto",
}


def _sanitize_table_name(name: str) -> str:
    if re.match(r"^[a-zA-Z][a-zA-Z0-9_]*$", name):
        return name
    out = re.sub(r"[^a-zA-Z0-9_]", "_", name)
    if not re.match(r"^[a-zA-Z]", out):
        out = "t_" + out
    out = re.sub(r"_+", "_", out)
    out = out.rstrip("_")
    return out


def _has_polars() -> bool:
    try:
        import polars as pl  # noqa: F401

        return True
    except ImportError:
        return False


def _convert_result(result: duckdb.DuckDBPyConnection) -> nw.DataFrame:
    """Convert a DuckDB result to a narwhals DataFrame, preferring polars if available."""
    if _has_polars():
        return nw.from_native(result.pl())
    return nw.from_native(result.df())


class PinSource(DataSource[nw.DataFrame]):
    """
    DataSource backed by a pin from a pins board.

    For parquet, CSV, JSON, and arrow pins, the cached files go straight into
    DuckDB (or through polars for arrow) without Python deserialization. Other
    pin types go through ``pin_read()`` first.

    After loading, DuckDB's external file access is locked down so that
    LLM-generated SQL cannot reach the filesystem.

    If the pin has a title, description, or tags,
    :class:`~querychat.QueryChat` uses them as the default
    ``data_description``, which you can override.

    Lazy queries with pins
    ~~~~~~~~~~~~~~~~~~~~~~

    ``PinSource`` materializes the full dataset into DuckDB. For large parquet
    pins where you want lazy query execution, read the pin files yourself and
    pass a Polars LazyFrame to :class:`~querychat.QueryChat` instead::

        paths = board.pin_download("my_pin")
        lf = pl.scan_parquet(paths[0])
        qc = QueryChat(lf, "my_pin")

    The pin files are still downloaded to a local cache ---
    ``pin_download()`` always fetches them. But rather than loading everything
    into memory, Polars reads the parquet file lazily.

    This approach skips the security lockdown that ``PinSource`` applies, so
    LLM-generated SQL can access files on the local system.
    """

    def __init__(
        self,
        board: Any,
        name: str,
        *,
        table_name: str | None = None,
        version: str | None = None,
    ):
        try:
            import pins  # noqa: F401
        except ImportError:
            raise ImportError(
                "The 'pins' package is required to use PinSource. "
                "Install it with: pip install querychat[pins]"
            ) from None

        effective_table_name = _sanitize_table_name(table_name or name)
        self.table_name = effective_table_name

        self._pin_meta_obj = board.pin_meta(name, version=version)
        pin_type = self._pin_meta_obj.type

        conn = duckdb.connect()
        try:
            if pin_type in DUCKDB_FILE_TYPES:
                paths = board.pin_download(name, version=version)
                if len(paths) != 1:
                    raise ValueError(
                        f"Pin '{name}' contains {len(paths)} files, but PinSource "
                        "requires a single-file pin (as created by pin_write())."
                    )
                reader_fn = DUCKDB_READER_FN[pin_type]
                if pin_type == "json":
                    conn.execute("INSTALL json")
                    conn.execute("LOAD json")
                conn.execute(
                    f'CREATE TABLE "{effective_table_name}" AS '
                    f"SELECT * FROM {reader_fn}(?)",
                    [paths[0]],
                )
            elif pin_type == "arrow" and _has_polars():
                # Arrow/IPC files can't be read natively by DuckDB, but
                # polars can read them directly — avoiding pin_read() overhead.
                import polars as pl

                paths = board.pin_download(name, version=version)
                if len(paths) != 1:
                    raise ValueError(
                        f"Pin '{name}' contains {len(paths)} files, but PinSource "
                        "requires a single-file pin (as created by pin_write())."
                    )
                arrow_df = pl.read_ipc(paths[0])
                vname = f"__pin_staging_{effective_table_name}"
                conn.register(vname, arrow_df)
                conn.execute(
                    f'CREATE TABLE "{effective_table_name}" AS '
                    f'SELECT * FROM "{vname}"'
                )
                conn.unregister(vname)
            else:
                import pandas as pd

                data = board.pin_read(name, version=version)
                if not isinstance(data, pd.DataFrame):
                    raise TypeError(
                        f"Pin '{name}' contains {type(data).__name__}, not a DataFrame. "
                        "PinSource requires the pin to contain a pandas DataFrame."
                    )
                vname = f"__pin_staging_{effective_table_name}"
                conn.register(vname, data)
                conn.execute(
                    f'CREATE TABLE "{effective_table_name}" AS '
                    f'SELECT * FROM "{vname}"'
                )
                conn.unregister(vname)

            duckdb_lock_down(conn)
        except Exception:
            conn.close()
            raise

        self._conn = conn

        # Store column names for validation
        result = self._conn.execute(f'SELECT * FROM "{effective_table_name}" LIMIT 0')
        self._colnames = [desc[0] for desc in result.description]

    def get_db_type(self) -> str:
        return "DuckDB"

    def get_schema(self, *, categorical_threshold: int) -> str:
        return duckdb_get_schema(self._conn, self.table_name, categorical_threshold)

    def execute_query(self, query: str) -> nw.DataFrame:
        check_query(query)
        return _convert_result(self._conn.execute(query))

    def test_query(
        self, query: str, *, require_all_columns: bool = False
    ) -> nw.DataFrame:
        check_query(query)
        normalized = query.rstrip().removesuffix(";")
        result = _convert_result(
            self._conn.execute(
                f"SELECT * FROM ({normalized}) AS subquery LIMIT 1"
            )
        )

        if require_all_columns:
            result_columns = set(result.columns)
            missing = set(self._colnames) - result_columns
            if missing:
                missing_list = ", ".join(f"'{c}'" for c in sorted(missing))
                original_list = ", ".join(f"'{c}'" for c in self._colnames)
                raise MissingColumnsError(
                    f"Query result missing required columns: {missing_list}. "
                    f"The query must return all original table columns. "
                    f"Original columns: {original_list}"
                )

        return result

    def get_data(self) -> nw.DataFrame:
        return _convert_result(
            self._conn.execute(f'SELECT * FROM "{self.table_name}"')
        )

    def cleanup(self) -> None:
        if self._conn:
            self._conn.close()

    @property
    def pin_meta(self) -> Any:
        return self._pin_meta_obj

    def get_data_description(self) -> str:
        meta = self._pin_meta_obj
        parts: list[str] = []
        if getattr(meta, "title", None):
            parts.append(meta.title)
        if getattr(meta, "description", None):
            parts.append(meta.description)
        tags = getattr(meta, "tags", None)
        if tags:
            parts.append(f"Tags: {', '.join(tags)}")
        return "\n\n".join(parts)
